File: tests/abi/edit_controller_layout.cpp
#include "pluginterfaces/vst/ivsteditcontroller.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Vst::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("kVstComponentControllerClass %s\n", kVstComponentControllerClass);
	std::printf ("ViewType.kEditor %s\n", Steinberg::Vst::ViewType::kEditor);
	std::printf ("ParameterInfo.kNoFlags %d\n", Steinberg::Vst::ParameterInfo::kNoFlags);
	std::printf ("ParameterInfo.kCanAutomate %d\n", Steinberg::Vst::ParameterInfo::kCanAutomate);
	std::printf ("ParameterInfo.kIsReadOnly %d\n", Steinberg::Vst::ParameterInfo::kIsReadOnly);
	std::printf ("ParameterInfo.kIsWrapAround %d\n", Steinberg::Vst::ParameterInfo::kIsWrapAround);
	std::printf ("ParameterInfo.kIsList %d\n", Steinberg::Vst::ParameterInfo::kIsList);
	std::printf ("ParameterInfo.kIsHidden %d\n", Steinberg::Vst::ParameterInfo::kIsHidden);
	std::printf ("ParameterInfo.kIsProgramChange %d\n", Steinberg::Vst::ParameterInfo::kIsProgramChange);
	std::printf ("ParameterInfo.kIsBypass %d\n", Steinberg::Vst::ParameterInfo::kIsBypass);
	std::printf ("RestartFlags.kReloadComponent %d\n", Steinberg::Vst::kReloadComponent);
	std::printf ("RestartFlags.kIoChanged %d\n", Steinberg::Vst::kIoChanged);
	std::printf ("RestartFlags.kParamValuesChanged %d\n", Steinberg::Vst::kParamValuesChanged);
	std::printf ("RestartFlags.kLatencyChanged %d\n", Steinberg::Vst::kLatencyChanged);
	std::printf ("RestartFlags.kParamTitlesChanged %d\n", Steinberg::Vst::kParamTitlesChanged);
	std::printf ("RestartFlags.kMidiCCAssignmentChanged %d\n", Steinberg::Vst::kMidiCCAssignmentChanged);
	std::printf ("RestartFlags.kNoteExpressionChanged %d\n", Steinberg::Vst::kNoteExpressionChanged);
	std::printf ("RestartFlags.kIoTitlesChanged %d\n", Steinberg::Vst::kIoTitlesChanged);
	std::printf ("RestartFlags.kPrefetchableSupportChanged %d\n", Steinberg::Vst::kPrefetchableSupportChanged);
	std::printf ("RestartFlags.kRoutingInfoChanged %d\n", Steinberg::Vst::kRoutingInfoChanged);
	std::printf ("RestartFlags.kKeyswitchChanged %d\n", Steinberg::Vst::kKeyswitchChanged);
	std::printf ("RestartFlags.kParamIDMappingChanged %d\n", Steinberg::Vst::kParamIDMappingChanged);
	std::printf ("ProgressType.AsyncStateRestoration %u\n", Steinberg::Vst::IProgress::AsyncStateRestoration);
	std::printf ("ProgressType.UIBackgroundTask %u\n", Steinberg::Vst::IProgress::UIBackgroundTask);
	std::printf ("KnobModes.kCircularMode %d\n", Steinberg::Vst::kCircularMode);
	std::printf ("KnobModes.kRelativCircularMode %d\n", Steinberg::Vst::kRelativCircularMode);
	std::printf ("KnobModes.kLinearMode %d\n", Steinberg::Vst::kLinearMode);

	PRINT_TYPE (ParameterInfo);
	PRINT_OFFSET (ParameterInfo, id);
	PRINT_OFFSET (ParameterInfo, title);
	PRINT_OFFSET (ParameterInfo, shortTitle);
	PRINT_OFFSET (ParameterInfo, units);
	PRINT_OFFSET (ParameterInfo, stepCount);
	PRINT_OFFSET (ParameterInfo, defaultNormalizedValue);
	PRINT_OFFSET (ParameterInfo, unitId);
	PRINT_OFFSET (ParameterInfo, flags);

	PRINT_TYPE (IComponentHandler);
	PRINT_TYPE (IComponentHandler2);
	PRINT_TYPE (IComponentHandlerBusActivation);
	PRINT_TYPE (IProgress);
	PRINT_TYPE (IEditController);
	PRINT_TYPE (IEditController2);
	PRINT_TYPE (IMidiMapping);
	PRINT_TYPE (IEditControllerHostEditing);
	PRINT_TYPE (IComponentHandlerSystemTime);
	print_iid ("IComponentHandler", Steinberg::Vst::IComponentHandler_iid);
	print_iid ("IComponentHandler2", Steinberg::Vst::IComponentHandler2_iid);
	print_iid ("IComponentHandlerBusActivation", Steinberg::Vst::IComponentHandlerBusActivation_iid);
	print_iid ("IProgress", Steinberg::Vst::IProgress_iid);
	print_iid ("IEditController", Steinberg::Vst::IEditController_iid);
	print_iid ("IEditController2", Steinberg::Vst::IEditController2_iid);
	print_iid ("IMidiMapping", Steinberg::Vst::IMidiMapping_iid);
	print_iid ("IEditControllerHostEditing", Steinberg::Vst::IEditControllerHostEditing_iid);
	print_iid ("IComponentHandlerSystemTime", Steinberg::Vst::IComponentHandlerSystemTime_iid);
	return 0;
}

File: tools/edit_controller_layout.zig
const std = @import("std");
const edit_controller = @import("vst3-zig").pluginterfaces.vst.ivsteditcontroller;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("kVstComponentControllerClass {s}\n", .{edit_controller.kVstComponentControllerClass});
    try stdout.print("ViewType.kEditor {s}\n", .{edit_controller.ViewType.kEditor});
    try stdout.print("ParameterInfo.kNoFlags {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kNoFlags});
    try stdout.print("ParameterInfo.kCanAutomate {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kCanAutomate});
    try stdout.print("ParameterInfo.kIsReadOnly {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kIsReadOnly});
    try stdout.print("ParameterInfo.kIsWrapAround {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kIsWrapAround});
    try stdout.print("ParameterInfo.kIsList {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kIsList});
    try stdout.print("ParameterInfo.kIsHidden {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kIsHidden});
    try stdout.print("ParameterInfo.kIsProgramChange {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kIsProgramChange});
    try stdout.print("ParameterInfo.kIsBypass {}\n", .{edit_controller.ParameterInfo.ParameterFlags.kIsBypass});
    try stdout.print("RestartFlags.kReloadComponent {}\n", .{edit_controller.RestartFlags.kReloadComponent});
    try stdout.print("RestartFlags.kIoChanged {}\n", .{edit_controller.RestartFlags.kIoChanged});
    try stdout.print("RestartFlags.kParamValuesChanged {}\n", .{edit_controller.RestartFlags.kParamValuesChanged});
    try stdout.print("RestartFlags.kLatencyChanged {}\n", .{edit_controller.RestartFlags.kLatencyChanged});
    try stdout.print("RestartFlags.kParamTitlesChanged {}\n", .{edit_controller.RestartFlags.kParamTitlesChanged});
    try stdout.print("RestartFlags.kMidiCCAssignmentChanged {}\n", .{edit_controller.RestartFlags.kMidiCCAssignmentChanged});
    try stdout.print("RestartFlags.kNoteExpressionChanged {}\n", .{edit_controller.RestartFlags.kNoteExpressionChanged});
    try stdout.print("RestartFlags.kIoTitlesChanged {}\n", .{edit_controller.RestartFlags.kIoTitlesChanged});
    try stdout.print("RestartFlags.kPrefetchableSupportChanged {}\n", .{edit_controller.RestartFlags.kPrefetchableSupportChanged});
    try stdout.print("RestartFlags.kRoutingInfoChanged {}\n", .{edit_controller.RestartFlags.kRoutingInfoChanged});
    try stdout.print("RestartFlags.kKeyswitchChanged {}\n", .{edit_controller.RestartFlags.kKeyswitchChanged});
    try stdout.print("RestartFlags.kParamIDMappingChanged {}\n", .{edit_controller.RestartFlags.kParamIDMappingChanged});
    try stdout.print("ProgressType.AsyncStateRestoration {}\n", .{@intFromEnum(edit_controller.ProgressType.AsyncStateRestoration)});
    try stdout.print("ProgressType.UIBackgroundTask {}\n", .{@intFromEnum(edit_controller.ProgressType.UIBackgroundTask)});
    try stdout.print("KnobModes.kCircularMode {}\n", .{@intFromEnum(edit_controller.KnobModes.kCircularMode)});
    try stdout.print("KnobModes.kRelativCircularMode {}\n", .{@intFromEnum(edit_controller.KnobModes.kRelativCircularMode)});
    try stdout.print("KnobModes.kLinearMode {}\n", .{@intFromEnum(edit_controller.KnobModes.kLinearMode)});

    try printType(stdout, "ParameterInfo", edit_controller.ParameterInfo);
    try printOffset(stdout, "ParameterInfo", "id", edit_controller.ParameterInfo, "id");
    try printOffset(stdout, "ParameterInfo", "title", edit_controller.ParameterInfo, "title");
    try printOffset(stdout, "ParameterInfo", "shortTitle", edit_controller.ParameterInfo, "shortTitle");
    try printOffset(stdout, "ParameterInfo", "units", edit_controller.ParameterInfo, "units");
    try printOffset(stdout, "ParameterInfo", "stepCount", edit_controller.ParameterInfo, "stepCount");
    try printOffset(stdout, "ParameterInfo", "defaultNormalizedValue", edit_controller.ParameterInfo, "defaultNormalizedValue");
    try printOffset(stdout, "ParameterInfo", "unitId", edit_controller.ParameterInfo, "unitId");
    try printOffset(stdout, "ParameterInfo", "flags", edit_controller.ParameterInfo, "flags");

    try printTuid(stdout, "IComponentHandler", edit_controller.icomponent_handler_iid);
    try printTuid(stdout, "IComponentHandler2", edit_controller.icomponent_handler2_iid);
    try printTuid(stdout, "IComponentHandlerBusActivation", edit_controller.icomponent_handler_bus_activation_iid);
    try printTuid(stdout, "IProgress", edit_controller.iprogress_iid);
    try printTuid(stdout, "IEditController", edit_controller.iedit_controller_iid);
    try printTuid(stdout, "IEditController2", edit_controller.iedit_controller2_iid);
    try printTuid(stdout, "IMidiMapping", edit_controller.imidi_mapping_iid);
    try printTuid(stdout, "IEditControllerHostEditing", edit_controller.iedit_controller_host_editing_iid);
    try printTuid(stdout, "IComponentHandlerSystemTime", edit_controller.icomponent_handler_system_time_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
